src/Simulation/Native/src/simulator/qir.hpp

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "qir_stdlib.h"
+#include <cstdint>
+
+#ifdef _WIN32
+#define QIR_EXPORT_API __declspec(dllexport)
+#else
+#define QIR_EXPORT_API
+#endif
+
+/*==============================================================================
+  Qubit & Result
+
+  These two types are opaque to the clients: clients cannot directly create, delete,
+  copy or check state of qubits and results. QUBIT* and RESULT* should never be
+  dereferenced in client's code.
+==============================================================================*/
+
+// Although QIR uses an opaque pointer to the type "QUBIT", it never points to an actual memory
+// and is never intended to be dereferenced anywhere - in the client code or in our runtime.
+// Runtime always operates in terms of qubit ids, which are integers. Qubit ids are casted
+// to this pointer type and stored as pointer values. This is done to ensure that qubit type
+// is a unique type in the QIR.
+
+class QUBIT;
+typedef intptr_t QubitIdType;
+
+class RESULT;
+
+struct QirRTuple
+{
+    PauliId pauli;
```

Comment:
Is the alignment / packing of this struct deterministic across platforms? More to the point, does the Rust side assume any specific layout for these structs?

Reply:
That's a great question, and no, currently there is no Rust code that uses this struct. This corresponds to the QIR generated from Q# for the argument tuple. From the Rust QIR stdlib standpoint, it's just an opaque tuple that's essentially a managed memory blob. So this needs to stay consistent with the Q# API and compiler QIR generation rather than Rust. I'll add comments on these to that effect, with pointers to the corresponding Q# signatures.